PLUGIN-1535 : Oracle decimal type Backward Compatibility Fix

MrRahulSharma opened the pull request:
Backward compatibility fix for Oracle number type with undefined precision and scale.
JIRA : https://cdap.atlassian.net/browse/PLUGIN-1535

MrRahulSharma committed:
Backward Compatibility Fix for Oracle Number type without precision and scale.

tivv reviewed (commented):
Please add a unit test for the fallback and refer respective JIRA in the PR title

MrRahulSharma commented:
> Please add a unit test for the fallback and refer respective JIRA in the PR title

Added Unit Test to validate if the schema is Decimal and does not match with the expected String type.
Updated JIRA in the description.

itsankit-google commented on oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSource.java:
```diff
@@ -153,6 +154,17 @@ public void validate(FailureCollector collector) {
+    @Override
+    protected void validateField(FailureCollector collector,
+                                 Schema.Field field, Schema actualFieldSchema, Schema expectedFieldSchema) {
+      // Validate if the field schema is of type Decimal and the expected is String
+      if (Schema.LogicalType.DECIMAL.equals(expectedFieldSchema.getLogicalType())
+          && actualFieldSchema.getType().equals(Schema.Type.STRING)) {
+        return;
```
The comment and the code does not match here: Comment says `expectedSchema` to be `String` and `actualSchema` but we are validating `actualSchema` to be `String`??

MrRahulSharma replied:
Updated

itsankit-google commented on oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSource.java:
```diff
@@ -153,6 +154,17 @@ public void validate(FailureCollector collector) {
+    protected void validateField(FailureCollector collector,
+                                 Schema.Field field, Schema actualFieldSchema, Schema expectedFieldSchema) {
+      // Validate if the field output schema is set to Decimal and the actual evaluated schema is String.
+      if (Schema.LogicalType.DECIMAL.equals(expectedFieldSchema.getLogicalType())
+          && actualFieldSchema.getType().equals(Schema.Type.STRING)) {
+        return;
+      }
```
Why is this change needed? When do we expect actual schema (schema in user's database) as `STRING` and output schema to be `DECIMAL`?

MrRahulSharma replied:
This change is needed to make sure that the pipeline upgrade continues to work post upgrade. Since the older handling of the precision less used to convert to the decimal type, and the new version would try to convert to the String type. In that case the output schema would contain Decimal(38, 0) (or something similar), and the code internally would try to identify the schema of the field(without precision and scale) as String.

albertshau replied:
would be good to put this explanation in the code comment

MrRahulSharma replied:
Added.

albertshau commented on oracle-plugin/src/test/java/io/cdap/plugin/oracle/OracleSourceDBRecordUnitTest.java:
```diff
@@ -95,6 +95,55 @@ public void validatePrecisionLessDecimalParsing() throws Exception {
+  /**
+   * Validate the precision less Numbers handling against following use cases.
+   * 1. Ensure that for Number(0,-127) non nullable type a String is returned by default.
+   * 2. Ensure that for Number(0,-127) non nullable type a String is returned,
```
this javadoc seems incorrect, should be returning a BigDecimal in the cases where the schema is defined as a Decimal.

MrRahulSharma replied:
Updated the comments

albertshau commented on oracle-plugin/src/test/java/io/cdap/plugin/oracle/OracleSourceDBRecordUnitTest.java:
```diff
@@ -95,6 +95,55 @@ public void validatePrecisionLessDecimalParsing() throws Exception {
 
+  /**
+   * Validate the precision less Numbers handling against following use cases.
+   * 1. Ensure that for Number(0,-127) non nullable type a String is returned by default.
```
I don't see this behavior in the test, there are only asserts for BigDecimal values returned.

MrRahulSharma replied:
Updated the comments.

albertshau commented on oracle-plugin/src/test/java/io/cdap/plugin/oracle/OracleSourceDBRecordUnitTest.java:
```diff
@@ -95,6 +95,55 @@ public void validatePrecisionLessDecimalParsing() throws Exception {
+    );
+
+    when(resultSet.getMetaData()).thenReturn(resultSetMetaData);
+    when(resultSet.getBigDecimal(eq(1), eq(0))).thenReturn(new BigDecimal("123"));
```
do we also have any tests where we actually read from an oracle DB? That would give more confidence than mocking the return value. For example, I see an OracleSourceTestRun, is that being executed?

MrRahulSharma replied:
OracleSourceTestRun does not execute.

fernst commented on oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSourceDBRecord.java:
```diff
@@ -223,8 +223,16 @@ private void handleOracleSpecificType(ResultSet resultSet, StructuredRecord.Buil
+              BigDecimal decimal = resultSet.getBigDecimal(columnIndex, getScale(field.getSchema()));
+              recordBuilder.setDecimal(field.getName(), decimal);
+            } else {
+              // In case of precision less decimal convert the field to String type
```
nit: "In case of a decimal with no precision/scale set, convert..."

fernst reviewed (commented):
Small nitpick

robinrielley commented on oracle-plugin/docs/Oracle-batchsink.md:
```diff
@@ -60,37 +60,38 @@ will be passed to the JDBC driver as connection arguments for JDBC drivers that
+    | VARCHAR2                       | string                |                                                            |
+    | NVARCHAR2                      | string                |                                                            |
+    | VARCHAR                        | string                |                                                            |
+    | NUMBER                         | string                | For NUMBER types defined without a precision and scale     |
```
add "defined" For NUMBER types without a defined precision and scale
